# Linear Classifier

Gegeben sind folgende Daten

<table><thead>
  <tr>
    <th></th>
    <th colspan="6">Training</th>
    <th colspan="2">Test</th>
  </tr></thead>
<tbody>
  <tr>
    <td>x1</td>
    <td>1</td>
    <td>-1</td>
    <td>0</td>
    <td>-1</td>
    <td>1</td>
    <td>0</td>
    <td>-1</td>
    <td>0.5</td>
  </tr>
  <tr>
    <td>x2</td>
    <td>-1</td>
    <td>0</td>
    <td>0</td>
    <td>2</td>
    <td>0</td>
    <td>1</td>
    <td>1</td>
    <td>-0.5</td>
  </tr>
  <tr>
    <td>y</td>
    <td>1</td>
    <td>1</td>
    <td>1</td>
    <td>-1</td>
    <td>-1</td>
    <td>-1</td>
    <td>1</td>
    <td>-1</td>
  </tr>
</tbody>
</table>

<br>
<img src="assets/s05_aufg1_plot.png" alt="Plot" style="width: 400px"/>

In [1]:
import numpy as np
np.set_printoptions(precision=4)

x1_training = np.array([1, -1, 0, -1, 1, 0])
x2_training = np.array([-1, 0, 0, 2, 0, 1])
y_training = np.array([1, 1, 1, -1, -1, -1])

x1_test = np.array([-1, 0.5])
x2_test = np.array([1, -0.5])
y_test = np.array([1, -1])

## Kleinste Quadrate Methode

> Berechnen Sie auf den Trainingsdaten mit der kleinste-Quadrate-Methode ein lineares Modell der Gestalt $p(x_1,x_2) = w_0 + w_1x_1 + w_2x_2$.

In [2]:
n = len(x1_training)
X = np.stack([np.ones(n), x1_training, x2_training]).T
print(X)

[[ 1.  1. -1.]
 [ 1. -1.  0.]
 [ 1.  0.  0.]
 [ 1. -1.  2.]
 [ 1.  1.  0.]
 [ 1.  0.  1.]]


Minimierung von $\Vert \vec{r} \Vert_2^2 = \Vert \vec{y} - \mathbf{X} \vec{w} \Vert \to \min{}$

In [3]:
W = np.linalg.lstsq(X, y_training, rcond=None)[0]
print(W)

[ 0.4324 -0.973  -1.2973]


In [15]:
def p(x):
    return W[0] + W[1] * x[0] + W[2] * x[1]

In [16]:
X_test = np.stack([x1_test, x2_test]).T

for i in range(len(y_test)):
    prediction = p(X_test[i])
    print(f'p({X_test[i]}) = {prediction}\t({y_test[i]})')

p([-1.  1.]) = 0.10810810810810811	(1)
p([ 0.5 -0.5]) = 0.594594594594595	(-1)


## Linearer Klassifikator

> Konstruieren Sie einen linearen Klassifikator aus dem Ergebnis von oben.
> 
> - Geben Sie die Entscheidungsgrenze an.
> - Welcher Anteil der Trainings- und Testdaten wird richtig klassifiziert?
> - Wie wird der neue Datenpunkt $\left(x_{1;neu}; x_{2;neu}\right) = \left(-0.5 ; 1.5\right)$ eingeteilt?

### Entscheidungsgrenzen

> Geben Sie die Entscheidungsgrenze an.

$$
\begin{aligned}
    0 &= p(\vec{x}) = w_0 + w_1x_1 + w_2x_2 \Rightarrow B \\
    \Rightarrow x_2 &= \frac{0.4324 - 0.973 x_1}{1.297} \\
    \Rightarrow x_2 &= -0.75 x_1 + \frac13
\end{aligned}
$$

### Klassifikation

> Welcher Anteil der Trainings- und Testdaten wird richtig klassifiziert?

In [ ]:
def y_hat(x):
    return np.sign(p(x))

In [40]:
X_training = np.stack([x1_training, x2_training]).T

for i in range(len(y_training)):
    prediction = y_hat(X_training[i])
    real = y_training[i]
    print(f'{i}: ', end='')
    if prediction == real:
        print('correct')
    else:
        print('false')

0: correct
1: correct
2: correct
3: correct
4: correct
5: correct


In [42]:
X_test = np.stack([x1_test, x2_test]).T

for i in range(len(y_test)):
    prediction = y_hat(X_test[i])
    real = y_test[i]
    print(f'{i}: ', end='')
    if prediction == real:
        print('correct')
    else:
        print('false')

0: correct
1: false


- Trainingsdaten werden alle korrekt eingeteilt
- Testdaten zur Hälfte
- Betrachtung des Plots $\to$ Test- *und* Traningsdaten sind nicht linear separierbar

### Neuer Datenpunkt

> Wie wird der neue Datenpunkt $\left(x_{1;neu}; x_{2;neu}\right) = \left(-0.5 ; 1.5\right)$ eingeteilt?

In [34]:
x_neu = [-0.5, 1.5]
np.sign(p(x_neu))

-1.0